### PR TITLE
ESLint consolidation

### PR DESCRIPTION
**Pull request by Antonio-Laguna**
### Description of the Change

This started as let's see what's available to do. Realized there were issues from dependabot that were failing Eslint and then giving an opinionated go to that.

These are the screenshots that showcase the issues:

Screenshot 2022-08-08 at 10 55 45

Screenshot 2022-08-08 at 10 55 54

Now a quick summary of what this does:

* Created a consolidated action that runs `lint` and `test` on every workspace that has it available.
* Added `lint` to every single package that didn't have it and also linted every file.
* Removed remnants of Node 12.
* Ensured license is consistent
* Separated tests on `eslint-config` to test each restricted syntax in isolation. The previous test had a bunch of `eslint-disable` modifiers plus tests only check that the files fail linting but not why. These files _only_ fail on the intended.
* Streamlined dependencies for `eslint-config`
* Removing husky and lint-staged: This goes in tandem with the `lint` in every package. Global `eslint` does seem like a good idea but not really since there's a mixture of packages with different linting needs. Each package now lints itself if available (all are now) and we get direct feedback on each PR from the linter. 
It was also not working correctly (on my tests) since either was too greedy that it picked `node_modules` or didn't lint enough files. Using this approach gave me a quick status check on every public package.

I **have not** added any changeset since I wanted to check this with the team first. I don't think there's a need to release anything as these are just chores. Though perhaps it'd be good to release `eslint-config` as `minor` to get the `peerDependency` updates.

### Checklist:

- [X] I have read the [**CONTRIBUTING**](https://github.com/10up/10up-toolkit/blob/develop/CONTRIBUTING.md) document.
- [X] My code follows the code style of this project.
- [X] My change requires a change to the documentation.
- [X] I have updated the documentation accordingly.
- [X] I have added tests to cover my change.
- [X] All new and existing tests passed.
- [ ] I have added a changeset to my PR. See [**CONTRIBUTING**](https://github.com/10up/10up-toolkit/blob/develop/CONTRIBUTING.md) document for instructions

### CODE REVIEW

**Comment by Antonio-Laguna** [comment]
The new tests are also good since not all features seem OK in every OS

**Comment by nicholasio** [comment]
Can we keep lint-staged and husky and have it trigger the individual tests on each package? I really like the idea of blocking commits if things fails to lint.

**Comment by nicholasio** [comment]
Did you check the failures on ubuntu and windows?

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+            is_base_os_version: true
+    steps:
+      - uses: actions/checkout@v3
+        with:
+          fetch-depth: 30
+      - uses: actions/setup-node@v3
+        with:
+          node-version: ${{ matrix.node }}
+
+      - name: env
```
Comment (nicholasio) [inline .github/workflows/test.yml:47]: What's this doing?

File: packages/eslint-config/package.json [hunk 2]
```diff
@@ -32,31 +32,20 @@
...
-    "eslint-plugin-jsx-a11y": "^6.5.1",
-    "eslint-plugin-prettier": "^4.0.0",
-    "eslint-plugin-react": "^7.27.0",
-    "eslint-plugin-react-hooks": "^4.3.0",
     "prettier": "^2.4.1"
   },
   "peerDependencies": {
     "@babel/core": "^7.16.0",
     "@babel/eslint-parser": "^7.16.3",
+	"@wordpress/eslint-plugin": "^12.0.0",
```
Comment (nicholasio) [inline packages/eslint-config/package.json:49]: Can we make this an optional dependency? We don't need it for non-wp projects and I'd rather not install WP specific things for non-wp projects;

File: .github/workflows/test.yml (hunk 1 above)
Comment (Antonio-Laguna) [reply .github/workflows/test.yml:47]: Building up env variables for the restore keys below

File: packages/eslint-config/package.json (hunk 2 above)
Comment (Antonio-Laguna) [reply packages/eslint-config/package.json:49]: We can, but we install JSX, React and React and React Hooks dependencies even if we don't use them. 

I think being inconsistent and also having mixed approaches depending on what you need to use felt a bit odd. This is a dev dependency in any case. 

Let me know your thoughts !

**Comment by Antonio-Laguna** [comment]
> Can we keep lint-staged and husky and have it trigger the individual tests on each package? I really like the idea of blocking commits if things fails to lint.

We could hook `lint-staged` to the whole linting but that feels odd. If I haven't changed those files I don't want to be blocked. At the same time, I think it makes sense for you to actually be able to commit things that aren't still 100% or that purposely fail to test things out (in any scenario).

Given you can't lint-stage per package's config (easily) I thought it makes the most sense to get a clean slate and be warned at a PR level so everyone in the PR process knows if something is suddenly not linting.  Especially when the packages had different linting issues that weren't caught before. 

In any case, this was meant as a discussion starter rather than "let's just roll with this".

>  Did you check the failures on ubuntu and windows?

I have, not sure yet what they are since the test seems to enter some sort of loop and has to be killed. Will investigate further.

File: packages/eslint-config/package.json (hunk 2 above)
Comment (nicholasio) [reply packages/eslint-config/package.json:49]: We don't install react, that's responsibility of the consumer. When using toolkit in WP context, the WP bundled react will be used.

toolkit already install `@wordpress/eslint-plugin` and everyone using eslint-config for WP would prob just use toolkit instead and would almost never install eslint-config directly, so leaving our eslint-config with no wp-specific stuff by default sounds reasonable to me.

Comment (Antonio-Laguna) [reply packages/eslint-config/package.json:49]: What I mean is that JSX and Hooks and React eslint packages are installed

Comment (Antonio-Laguna) [reply packages/eslint-config/package.json:49]: IMHO it's a bit strange to be asked to install other packages. Adding to that is the fact that different linting scenarios are treated differently which is why I did that.

**Comment by nicholasio** [comment]
> We could hook `lint-staged` to the whole linting but that feels odd. If I haven't changed those files I don't want to be blocked. At the same time, I think it makes sense for you to actually be able to commit things that aren't still 100% or that purposely fail to test things out (in any scenario).

lint-staged would only lint the changed files.

> Given you can't lint-stage per package's config (easily) I thought it makes the most sense to get a clean slate and be warned at a PR level so everyone in the PR process knows if something is suddenly not linting. Especially when the packages had different linting issues that weren't caught before.

It's a valid approach but we've been historically using husky and lint-staged to get linting feedback at commit level. One downside of having to wait for PR bots is that it might take a while, with lint-staged you can near instant feedback at commit time and you avoid linting back-and-forth in the PR.

**Comment by Antonio-Laguna** [comment]
> lint-staged would only lint the changed files.

Yes but it's not able to apply different linting scenarios per file which is why I changed. That's the main reason I changed that.

> It's a valid approach but we've been historically using husky and lint-staged to get linting feedback at commit level. One downside of having to wait for PR bots is that it might take a while, with lint-staged you can near instant feedback at commit time and you avoid linting back-and-forth in the PR.

It's also a valid approach to use husky and lint-staged. I'm not saying it isn't and if it wasn't because it's not doable to have lint-staged with different scenarios (global eslint wasn't working) I wouldn't have changed it. Also, we still got unlinted files somehow within the repo which, somehow, escaped that instant feedback at commit time.

In any case, if you can think of other solutions let me know as I'm more than happy to take another path. I feel like lint-staged is not the right fit for monorepos with different needs and different packages but I could be wrong.

**Comment by nicholasio** [comment]
> In any case, if you can think of other solutions let me know as I'm more than happy to take another path. I feel like lint-staged is not the right fit for monorepos with different needs and different packages but I could be wrong.

fair enough.

Looks like the test on windows is still failling/running indefinatelly, once that's resolved I think we should be good to merge this.

**Comment by Antonio-Laguna** [comment]
I’m not sure if the tests will pass, will try to, at least, make them fail. I got a virtual box instance of W11 and was able to move a bit forward but not having windows makes it complicated 😅

Will keep you posted!

**Comment by Antonio-Laguna** [comment]
Windows is failing because of this issue https://github.com/npm/cli/issues/5454 

I'll try to come up with some workaround :/

**Comment by Antonio-Laguna** [comment]
After 40 commits, countless hours, and (20gb) windows virtual machine this seems to be all green now.

cc @fabiankaegy @nicholasio

File: packages/eslint-config/package.json [hunk 3]
```diff
@@ -2,65 +2,62 @@
...
-    "test:wordpress": "node fixtures/test-eslint-config --config-file=./wordpress.js",
-    "test:node": "node fixtures/test-eslint-config --config-file=./node.js"
+	"lint": "eslint .",
+	"test": "node fixtures/test-eslint-config.js",
+	"test:index": "node fixtures/test-eslint-config.js --config-file=index",
+	"test:react": "node fixtures/test-eslint-config.js --config-file=react",
+	"test:wordpress": "node fixtures/test-eslint-config.js --config-file=wordpress",
+	"test:node": "node fixtures/test-eslint-config.js --config-file=nodejs"
+  },
+  "exports": {
```
Comment (nicholasio) [inline packages/eslint-config/package.json:16]: Would this require a major bump?

Comment (Antonio-Laguna) [reply packages/eslint-config/package.json:16]: No.

This was a major concern. I didn't want to create a breaking change. Renaming node to nodejs would have been breaking so I tried with exports to see if it worked. 

The exports change just required changes on some internal routes but no changes on external package usages.

**Review by nicholasio — APPROVED** [review]
@Antonio-Laguna I made one small update to make `@wordpress/eslint-plugin` optional. I did that since we also use the eslint-config on non-wp projects and it's how it was set up from the start. I agree we can make the other deps non-optional. 

For wp projects we use toolklit which already install the eslint-plugin.

Thanks for all your work on this!

**Comment by Antonio-Laguna** [comment]
That's a sensible addition. Thanks for doing that :)

**Pull request merged**